hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataAvroResolver.java

```diff
@@ -99,19 +104,24 @@ public Stream<MappingField> resolveAndValidateFields(
         if (userFields.isEmpty()) {
             throw QueryException.error("Column list is required for Avro format");
         }
-        String inlineSchema = options.get(isKey ? OPTION_KEY_AVRO_SCHEMA : OPTION_VALUE_AVRO_SCHEMA);
-        if (inlineSchema != null && options.containsKey("schema.registry.url")) {
+        Map<QueryPath, MappingField> fieldsByPath = extractFields(userFields, isKey);
+
+        Schema schema = getSchemaId(fieldsByPath, schemaJson -> {
+            // HazelcastKafkaAvro[De]Serializer obtains the schema from mapping options
```

Comment:
in what case do we need mutable `Mapping.options()`? Is the only case the legacy one, when `jet.sql` schema is generated from mapping fields? But if so, don't we use default Kafka Avro Ser/De instead of HazelcastKafkaAvro[De]Serializer in such case?

This is also very poorly maintainable. Note that options are passed through some objects (eg. `SqlExternalResource`) and methods before reaching here, and each of them must not only have mutable options, but also the `options` must be reference to internal representation of `Mapping`, not a copy (even mutable).

Reply:
`getSchemaId()` prefers the schema defined by `__key`/`this` fields if there is any, and this preference should be propagated to `HazelcastKafkaAvro[De]Serializer` somehow. As you pointed out, the current propagation method is very fragile. I actually tried to come up with a proper implementation, but it quickly became an invasive change. I will give it another try.

Reply:
I like the new version much better. Do I understand correctly that this is needed to override schema from Mapping OPTIONS by schema from UDT OPTIONS if such UDT is used as entire key or value? If so, in such Avro schema on mapping level will be silently ignored which might be puzzling (mutable options were even more surprising), but I do not know if we can do anything about it apart from dedicated validation for such case, so we can leave it as is.

Reply:
> Do I understand correctly that this is needed to override schema from Mapping OPTIONS by schema from UDT OPTIONS if such UDT is used as entire key or value?

Correct.
> Avro schema on mapping level will be silently ignored which might be puzzling

This is not a bug, but a feature. As long as we document the behavior, there shouldn't be any problem.